### PR TITLE
feat(platform/azure): implement automergeStrategy for Azure DevOps platform

### LINKED ISSUE #7310: Possibility to set mergeType when using azureAutoComplete
I would like to be able to set the mergeType when setting azureAutoComplete.
Right now the setting defaults to Merge (no fast forward) which can conflict with defined PR policies.

image

**Did you already have any implementation ideas?**

Add a property azureMergeType with the possible values Merge, SquashCommit, RebaseAndFastForward, SemiLinearMerge.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

As workaround I have to manualy complete the PR by clicking Complete and selecting the correct merge type. See screenshot.

### LINKED ISSUE #10851: Implement the explicit `automergeStrategy` for the Azure DevOps platform.
**What would you like Renovate to be able to do?**

The `automergeStrategy` configuration item introduced in #10627 should be implemented for the Azure DevOps platform, allowing users to configure the explicit merge strategy used by Renovate when automerging pull requests.

**Did you already have any implementation ideas?**

The optional `MergeStrategy` argument can be provided to Azure's `mergePr` implementation. The current implementation queries the target to see what the configured merge strategy is, but the Renovate-specific strings for automergeStrategy can be mapped to Azure-specific strings when submitting the merge request to Azure.

### PR DESCRIPTION
**Pull request by ghost**
## Changes

This PR allows users to define the merge strategy to use when automerging PRs.

Fixes #7310.
Fixes #10851.
Contributes to #16884 (we shouldn't close this issue as there are still some TODO in the code to be fixed).

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/azure/index.ts [hunk 1]
```diff
@@ -461,6 +462,22 @@ async function getMergeStrategy(
   );
 }
 
+function mapMergeStrategy(
+  mergeStrategy?: MergeStrategy,
+): GitPullRequestMergeStrategy {
+  switch (mergeStrategy) {
+    case 'rebase':
+    case 'fast-forward':
```
Comment (joegoldman2) [inline lib/modules/platform/azure/index.ts:470]: I'm not 100% sure about this mapping (fast-forward => rebase) but as the documentation explains, the only way to perform a fast-forward merge  on Azure DevOps is to rebase: https://learn.microsoft.com/en-us/azure/devops/repos/git/complete-pull-requests?view=azure-devops&tabs=browser#complete-a-pull-request

**Comment by joegoldman2** [comment]
This PR is ready to be reviewed.

File: lib/modules/platform/azure/index.ts [hunk 2]
```diff
@@ -461,6 +462,22 @@ async function getMergeStrategy(
...
+      return GitPullRequestMergeStrategy.Rebase;
+    case 'merge-commit':
+      return GitPullRequestMergeStrategy.NoFastForward;
+    case 'squash':
+      return GitPullRequestMergeStrategy.Squash;
+    default:
+      return GitPullRequestMergeStrategy.NoFastForward;
```
Comment (joegoldman2) [inline lib/modules/platform/azure/index.ts:477]: This default value is the same as here (used when `automergeStrategy` is `auto`):
https://github.com/renovatebot/renovate/blob/b8a28e0309def1ede5e723dbb460fde52e520892/lib/modules/platform/azure/azure-helper.ts#L164-L176

**Comment by joegoldman2** [comment]
This PR is ready to be reviewed.

**Comment by joegoldman2** [comment]
This PR is ready to be reviewed.

File: lib/modules/platform/azure/index.ts (hunk 1 above)
Comment (joegoldman2) [reply lib/modules/platform/azure/index.ts:470]: cc @JamieMagee

File: lib/modules/platform/azure/index.spec.ts [hunk 3]
```diff
@@ -1010,12 +1013,125 @@ describe('modules/platform/azure/index', () => {
...
+            targetBranch: 'dev',
+            prTitle: 'The Title',
+            prBody: 'Hello world',
+            labels: ['deps', 'renovate'],
+            platformOptions: {
+              automergeStrategy,
+              usePlatformAutomerge: true,
+            },
+          });
+          expect(pr).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/platform/azure/index.spec.ts:1127]: no new snapshots, use toEqual or toMatchObject. read our dev guide

File: lib/modules/platform/azure/index.ts (hunk 1 above)
Comment (viceice) [reply lib/modules/platform/azure/index.ts:470]: auto merge will only be done when branch is up to date, so git will simply push the commit to the target branch.

File: lib/modules/platform/azure/index.ts [hunk 4]
```diff
@@ -461,6 +462,22 @@ async function getMergeStrategy(
   );
 }
 
+function mapMergeStrategy(
```
Comment (viceice) [inline lib/modules/platform/azure/index.ts:465]: please move to commonts or util.ts, whatever exists

File: lib/modules/platform/azure/index.spec.ts (hunk 3 above)
Comment (joegoldman2) [reply lib/modules/platform/azure/index.spec.ts]: done, snapshots removed.

File: lib/modules/platform/azure/index.ts (hunk 4 above)
Comment (joegoldman2) [reply lib/modules/platform/azure/index.ts]: Done, moved to util.ts.

**Review by viceice — APPROVED** [review]

**Review by secustor — APPROVED** [review]

**Pull request merged**